Source/WebCore/Modules/webcodecs/WebCodecsVideoFrameAlgorithms.cpp

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebCodecsVideoFrameAlgorithms.h"
+
+#if ENABLE(WEB_CODECS)
+
+#include "DOMRectReadOnly.h"
+#include "VideoColorSpace.h"
+
+namespace WebCore {
+
+// https://w3c.github.io/webcodecs/#valid-videoframebufferinit
+bool isValidVideoFrameBufferInit(const WebCodecsVideoFrame::BufferInit& init)
+{
+    if (!init.codedWidth || !init.codedHeight)
+        return false;
+
+    if (init.visibleRect) {
+        if (init.visibleRect->x < 0 || init.visibleRect->y < 0 || init.visibleRect->width < 0 || init.visibleRect->height < 0)
+            return false;
+        if (!std::isfinite(init.visibleRect->x) || !std::isfinite(init.visibleRect->y) || !std::isfinite(init.visibleRect->width) || !std::isfinite(init.visibleRect->height))
+            return false;
+
+        if (init.visibleRect->y + init.visibleRect->height > init.codedHeight)
+            return false;
+
+        if (init.visibleRect->x + init.visibleRect->width > init.codedWidth)
+            return false;
+    }
+
+    if (init.displayWidth && !init.displayWidth)
+        return false;
+    if (init.displayHeight && !init.displayHeight)
+        return false;
+    return true;
+}
+
+// https://w3c.github.io/webcodecs/#videoframe-verify-rect-offset-alignment
+bool verifyRectOffsetAlignment(VideoPixelFormat format, const DOMRectInit& rect)
+{
+    switch (format) {
+    case VideoPixelFormat::I420:
+    case VideoPixelFormat::I420A:
+        return !(static_cast<unsigned>(rect.x) % 2) && !(static_cast<unsigned>(rect.y) % 2);
```

Comment:
It would be nice to have a function that does this instead of repeating the checks inline over and over.

Reply:
Will change the ordering of the enum and will add a isMultiple function.